File: portfolio-api/src/clients/frankfurter.ts
import { z } from "zod";

const FrankfurterResponseSchema = z.object({
  amount: z.number(),
  base: z.string(),
  date: z.string(),
  rates: z.record(z.string(), z.number()),
});

export class FrankfurterClient {
  private readonly baseUrl = "https://api.frankfurter.app";

  /** Returns the EUR/USD rate for a given date (USD per 1 EUR) */
  async getRate(date: string): Promise<number> {
    const url = `${this.baseUrl}/${date}?from=EUR&to=USD`;
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Frankfurter API error: ${response.status} ${response.statusText}`);
    }
    const raw: unknown = await response.json();
    const parsed = FrankfurterResponseSchema.parse(raw);
    const usd = parsed.rates["USD"];
    if (usd === undefined) {
      throw new Error(`No USD rate found in Frankfurter response for ${date}`);
    }
    return usd;
  }
}

File: fastapi/app/clients/frankfurter.py
import httpx


class FrankfurterClient:
    # api.frankfurter.app 301-redirects to frankfurter.dev. httpx does not follow
    # redirects unless asked, so without `follow_redirects` every rate lookup
    # raised and silently fell back to a hardcoded 1.1.
    BASE_URL = "https://api.frankfurter.dev/v1"

    async def get_rate(self, date: str) -> float:
        url = f"{self.BASE_URL}/{date}?base=EUR&symbols=USD"
        async with httpx.AsyncClient(timeout=10, follow_redirects=True) as client:
            r = await client.get(url)
        r.raise_for_status()
        data = r.json()
        usd = data.get("rates", {}).get("USD")
        if usd is None:
            raise ValueError(f"No USD rate for {date}")
        return float(usd)
